cupy/core/_routines_indexing.pyx

```diff
@@ -412,23 +424,44 @@ cdef ndarray _simple_getitem(ndarray a, list slice_list):
     return v
 
 
+_get_index_preamble = '''
+__device__ S get_index(const S* index, bool imcomplete_index, int blk_size,
+                       long long i, long long size) {
+    S j = index[i] - 1;
+    if (imcomplete_index && i < size - 1) {
+        long long ii = i % blk_size;
+        long long bi = i - ii;
+        if (ii < blk_size - 1 && bi > 0) {
+            j += index[bi - 1];
+        }
+    }
+    return j;
+}'''
+
+
 _nonzero_kernel_1d = ElementwiseKernel(
-    'T src, S index', 'raw U dst',
-    'if (src != 0) dst[index - 1] = i',
-    'nonzero_kernel_1d')
+    'T src, raw S index, bool imcomplete_index, int32 blk_size', 'raw U dst',
```

Comment:
```suggestion
    'T src, raw S index, bool incomplete_index, int32 blk_size', 'raw U dst',
```
😄

Reply:
As for the spelling of "incomplete", I'm consistently wrong! 😅